src/btree/bt_sync.c

```diff
@@ -122,13 +122,12 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
 static int
 __sync_ref_mark_deleted(WT_SESSION_IMPL *session, WT_REF *ref)
 {
-    /* Ignore root pages as they are never be deleted */
+    /* Ignore root pages as they can never be deleted. */
     if (__wt_ref_is_root(ref))
         return (0);
 
-    /* Guard against marking a page to be deleted when its reconciliation state is not empty. */
-    if (ref->page != NULL && ref->page->modify != NULL &&
-      ref->page->modify->rec_result != WT_PM_REC_EMPTY)
+    /* Guard against marking an in-memory page to be deleted. */
+    if (ref->page != NULL)
```

Comment:
I'd go with `WT_REF.state == WT_REF_DISK`. You want something like this:

```
if (ref->state != WT_REF_DISK || !WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
        return (0);
```

Otherwise, you can do the test, find `WT_REF.state == WT_REF_DISK`, then swap it to `WT_REF_LOCKED`, but the page went to `WT_REF_MEM` between the test and the lock.

Reply:
Thanks, Fixed accordingly.